### PR TITLE
Fix Sodium 0.5 incompatibility

### PR DESCRIPTION
**Pull request by BluSpring**
Issue: #175

In Sodium 0.5, they appear to have moved from Vanilla's LightmapTextureManager to using [LightDataAccess](https://github.com/CaffeineMC/sodium-fabric/blob/mc1.20.1-0.5.0/src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java). However, this brings in the caveat of not using Vanilla's `WorldRenderer#getLightmapCoordinates`, which LambDynLights relies on to provide dynamic light data to the renderer.

This PR creates a mixin into LightDataAccess that passes the dynamic light data onto Sodium. This unfortunately looks like the only way to both provide this lighting data and make it limited to the renderer alone. An injection into `BlockRenderView#getLightLevel` may also be plausible to provide that light data, but it would disrupt the world's light data for both Vanilla and mods.

**WARNING**: I observed large amounts of chunk flickering on my integrated GPU after making this change, so I suspect that more work will have to be done for this change to actually be viable.

### CODE REVIEW

File: src/main/java/dev/lambdaurora/lambdynlights/LambDynLightsCompat.java [hunk 1]
```diff
@@ -47,4 +47,9 @@ public static boolean isSodium010Installed() {
 		return FabricLoader.getInstance().getModContainer("sodium").map(mod -> mod.getMetadata().getVersion().getFriendlyString().startsWith("0.1.0"))
 				.orElse(false);
 	}
+
+	public static boolean isSodium05XInstalled() {
+		return FabricLoader.getInstance().getModContainer("sodium").map(mod -> mod.getMetadata().getVersion().getFriendlyString().startsWith("0.5."))
```
Comment (LambdAurora) [inline src/main/java/dev/lambdaurora/lambdynlights/LambDynLightsCompat.java]: This will easily break once Sodium releases a 0.6 version

**Comment by BluSpring** [comment]
i accidentally hit commit & push before realizing a better idea-

**Comment by Kichura** [comment]
Decided to give it a play-test on my radeon dGPU + iGPU and i cannot seem to reproduce the flickering chunks, is it intel-related?

**Comment by BluSpring** [comment]
> Decided to give it a play-test on my radeon dGPU + iGPU and i cannot seem to reproduce the flickering chunks, is it intel-related?

I don't use Intel, the flickering chunks were on my Radeon Vega 8. Though this *was* in debugging mode in a development environment, so that might also change things.

**Comment by Kichura** [comment]
> > Decided to give it a play-test on my radeon dGPU + iGPU and i cannot seem to reproduce the flickering chunks, is it intel-related?
> 
> I don't use Intel, the flickering chunks were on my Radeon Vega 8. Though this _was_ in debugging mode in a development environment, so that might also change things.

Possibly although i will attempt to run a second play-test using an Intel iGPU instead to see what comes out.

**Comment by Kichura** [comment]
> > > Decided to give it a play-test on my radeon dGPU + iGPU and i cannot seem to reproduce the flickering chunks, is it intel-related?
> > 
> > 
> > I don't use Intel, the flickering chunks were on my Radeon Vega 8. Though this _was_ in debugging mode in a development environment, so that might also change things.
> 
> Possibly although i will attempt to run a second play-test using an Intel iGPU instead to see what comes out.

Did the play-test and i still can't note flickerings from moving around with an torch although there are minor delays in certain places although hard to spot.

**Comment by caproven** [comment]
As an avid fan & user of the mod, thank you for making a PR! I tested this myself (AMD Ryzen cpu + Nvidia gpu) and didn't observe any flickering. Dynamic lights don't seem as "fluid" as before when the subject is moving, giving a subtle stuttering effect. This was mostly observed with flame arrows but is noticeable for common scenarios like the player holding a torch.

**Comment by joestr** [comment]
> As an avid fan & user of the mod, thank you for making a PR! I tested this myself (AMD Ryzen cpu + Nvidia gpu) and didn't observe any flickering. Dynamic lights don't seem as "fluid" as before when the subject is moving, giving a subtle stuttering effect. This was mostly observed with flame arrows but is noticeable for common scenarios like the player holding a torch.

Can confirm. No flickering but the lighting is not that fluid anymore.

**Comment by Energobro** [comment]
where to get the file?

**Comment by Energobro** [comment]
There is flickering, rtx 3050 ti

**Comment by LambdAurora** [comment]
Please run `./gradlew applyLicenses` and commit to fix the build.

File: src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/LightDataAccessMixin.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+public class LightDataAccessMixin {
+    // In Sodium 0.5, LightDataAccess has replaced the Vanilla LightmapTextureManager,
+    // and no longer uses WorldRenderer#getLightmapCoordinates.
+    // Unfortunately, this breaks LambDynLights, and as a result is forced to mixin to Sodium.
+    @Dynamic
+    @Redirect(method = "compute", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/BlockRenderView;getLightLevel(Lnet/minecraft/world/LightType;Lnet/minecraft/util/math/BlockPos;)I", ordinal = 0, remap = true), remap = false)
+    private int lambdynlights$modifyBlockLight(BlockRenderView world, LightType type, BlockPos pos, int x, int y, int z) {
+        int original = world.getLightLevel(type, pos);
+
+        if (!world.getBlockState(pos).isOpaqueFullCube(world, pos) && LambDynLights.get().config.getDynamicLightsMode().isEnabled()) {
```
Comment (Technici4n) [inline src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/LightDataAccessMixin.java]: This `!world.getBlockState(pos).isOpaqueFullCube(world, pos)` check is somewhat inefficient - Sodium already computed it before: https://github.com/CaffeineMC/sodium-fabric/blob/21e4e6d2f09f8d6b8ae7e5a5fab4a13c1f8379ce/src/main/java/me/jellysquid/mods/sodium/client/model/light/data/LightDataAccess.java#L68.

Comment (BluSpring) [reply src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/LightDataAccessMixin.java]: i'm aware of this, but i couldn't manage to figure out how to recapture the locals without needing MixinExtra.

Comment (BluSpring) [reply src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/LightDataAccessMixin.java]: you can actually see remnants of my feeble attempt of trying to capture locals from the `int x, int y, int z` params that I forgot to remove

**Comment by BluSpring** [comment]
> Please run `./gradlew applyLicenses` and commit to fix the build.

done

File: src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/LightDataAccessMixin.java (hunk 2 above)
Comment (Technici4n) [reply src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/LightDataAccessMixin.java]: Ah yes, it seems that redirect cannot capture local variables. Oh well.

**Comment by TheRealJake12** [comment]
yeah I'm getting delayed lighting as well. Also the lighting looks different? Idk.
Intel Iris XE integrated gpu btw
https://github.com/LambdAurora/LambDynamicLights/assets/84357907/320efe56-0bea-48ff-ac0c-429529890841

**Comment by NotDarkn** [comment]
currently using your build of LambDynamicLights and it does seem to work with Sodium 0.5, however the only issue I've found atm is that it's not as fluid as it should be. No flickering though, and I'm on an iGPU 6000.

**Comment by BluSpring** [comment]
seems a very common factor is the update fluidity. making a RETURN injection into ArrayLightDataCache and/or HashLightDataCache's `get` methods instead of `LightDataAccess#compute` might be able to resolve this.. i'm gonna go try that.

**Comment by BluSpring** [comment]
I'm unable to properly verify if this commit will improve light update fluidity, can someone help verify? An outstanding problem that I am noticing with light update fluidity is that the sub-chunk (16x16x16) below you will update slower compared to your current sub-chunk, this appears to be a little quirk with Sodium 0.5 though. I'm not sure how that can be dealt with.

**Comment by TheRealJake12** [comment]
> I'm unable to properly verify if this commit will improve light update fluidity, can someone help verify? An outstanding problem that I am noticing with light update fluidity is that the sub-chunk (16x16x16) below you will update slower compared to your current sub-chunk, this appears to be a little quirk with Sodium 0.5 though. I'm not sure how that can be dealt with.

gonna try it in a few hours gimme a lil bit

**Comment by TheRealJake12** [comment]
> > I'm unable to properly verify if this commit will improve light update fluidity, can someone help verify? An outstanding problem that I am noticing with light update fluidity is that the sub-chunk (16x16x16) below you will update slower compared to your current sub-chunk, this appears to be a little quirk with Sodium 0.5 though. I'm not sure how that can be dealt with.
>
> gonna try it in a few hours gimme a lil bit

lighting still feels a little delayed but is better I think. The light still looks weird though.
https://github.com/LambdAurora/LambDynamicLights/assets/84357907/0119c7bd-85ce-4005-8624-e2ed7c7c9229

**Comment by LambdAurora** [comment]
I have tested this PR, and I am noticing significant delays in light propagation, and the light pattern is also *very strange*.
Someone might need to do further research and maybe find another injection point, or at least the cause of the weirdness.

**Comment by MUKSC** [comment]
The light delay issue was introduced in [this](https://github.com/CaffeineMC/sodium-fabric/commit/af282d1771d5ede4fc813a1f630de64722f34a8c) commit, and the strange light pattern issue was introduced in [this](https://github.com/CaffeineMC/sodium-fabric/commit/dfb933baeb3cf94e39ecf50a53d2a600ffc1d37f) commit, if that helps.

**Comment by MUKSC** [comment]
I realized that the light delay issue is actually a precision issue.
In the original behavior, it tries to save more decimals as you can see [here](https://github.com/LambdAurora/LambDynamicLights/blob/1.20/src/main/java/dev/lambdaurora/lambdynlights/LambDynLights.java#L168).
But in this PR, it's casting double to int in [here](https://github.com/BluSpring/LambDynamicLights/blob/1.20/src/main/java/dev/lambdaurora/lambdynlights/mixin/sodium/ArrayLightDataCacheMixin.java#L45) to replace block light, so it loses all decimals here.

I think what we can do is either:
1. Inject to `ArrayLightDataCache.get` and save block position, then inject to `LightDataAccess.getLightMap` and use saved block position.
2. Redirect all calls to `LightDataAccess.getLightMap/getEmissiveLightmap` and use captured local variables to know block position.

Method 1 is hacky way and easily break with something like this:
```java
int word1 = cache.get(x1, y1, z1);
int word2 = cache.get(x2, y2, z2);
int lm1 = getLightmap(word1);
int lm2 = getLightmap(word2);
```
Method 2 is more safer way, but it would be a complete mess.

Both methods are pretty bad, so it might be better to send PR to Sodium to address this.
A potential solution could be to modify `LightDataAccess` to cache the lightmap as well.

**Comment by redmine4404** [comment]
i don't really understand how all of this work but I have the impression it's a sync issue, it's more visible when turn off smooth lightning : 

https://github.com/LambdAurora/LambDynamicLights/assets/96315018/9a892d5f-9f90-41ac-807b-8139e8790c14

**Comment by donington** [comment]
I suspect the weird lighting is because luminance is not being updated.  The original lightmap MUKSC referenced above was updating both the block light and luminance.

I pushed a pull request to BluSpring a couple of days ago since my git is rusty and I didn't want to clutter things with a failed attempt.  Which I did fail and just fixed thanks to UnablePath noticing my failed push attempt.

I have [a branch](https://github.com/donington/LambDynamicLights) with an updated lighting that utilizes Sodium's system a bit further.  I can't seem to post a test release properly because I haven't remembered how to add a tag properly yet, but the code compiles and looks a lot smoother on my end.  If anyone who is able to compile can test that would be great as I'd love to see this fixed on the latest version.

**Comment by BluSpring** [comment]
In my previous commit, I had just noticed something particularly interesting: The flat lighting data did *not* match the data without Sodium.

Without Sodium:
image

With Sodium + my PR:
image


In @donington's branch, I had noticed these things on my end:
- The light update fluidity was slightly improved, but not by much.
- The strange light patterns were still occurring.
- The flat lighting data was still very different from the data without Sodium.

In my latest commit, I decided to try reading further in understanding how Sodium handled lighting in both pipelines, and realized something: It still uses `LightmapTextureManager#pack` in some places.
This made me realize that LambDynLights' `getLightmapWithDynamicLight` method could still be used, and would hopefully solve some issues with the light fluidity.
After some time moving things around and testing, I finally got those changes functioning, with the flat lighting data to be matching again.

image

There is however one final problem: The smooth lighting data still generates some weird patterns, and I'm currently unsure as to what is causing it.
image

However, from what I can tell, the light updating seems to be much smoother now, at least on my side.

**Comment by LambdAurora** [comment]
Now that I'm seeing the top-down view of the weird light pattern, it's genuinely fun to see? I think at this point it's not bothering me as much if the delay issue is fixed.

Could you please undo the copyright header changes for previously existing files? The dates are supposed to be the dates of the creation of the file, and the gradle plugin seems to have gotten confused and not recognized the files.

**Comment by BluSpring** [comment]
Done. When I ran the `applyLicenses` task (I'm on Windows), the copyright symbol didn't seem to be added correctly either, and I had to fix it by manually doing a Replace All in VS Code, which is why there were two commits for it.

**Review by LambdAurora — APPROVED** [review]
I feel like this PR has reached a good enough stage for me to merge it in in a couple of hours.

I might rework a little bit some details to loosen up the requirement to have Sodium 0.5 for compilation to work through the power of aggressive Mixin Pseudo.

Otherwise, congratulations on the PR and huge thanks to everyone who contributed to it, I don't think I would have had the courage to figure out this solution!

**Comment by MUKSC** [comment]
Nice!
Btw I'll leave here a potential cause of light pattern issue.
Maybe we can fully fix the issue once this is fixed.
https://github.com/CaffeineMC/sodium-fabric/issues/2004

**Comment by embeddedt** [comment]
> Btw I'll leave here a potential cause of light pattern issue.
Maybe we can fully fix the issue once this is fixed.

Drive-by comment, as I am not familiar with the LDL code, but if it previously assumed it could use all 8 bits of the lightmap coordinates, that is no longer the case in Sodium 0.5 - you can only use the top 4. Providing something like `0xef` for the coordinate will result in Sodium interpreting it as `0xe` instead of `0xf`, causing dark patches.

**Comment by BluSpring** [comment]
LDL does in fact use all 8 bits, and the bottom 4 being unusable was what I had assumed, and that does appear to be the case with the smooth lighting pipeline, hence why I switched to just directly modifying the lightmap value returns, but what I do find interesting is that the flat lighting pipeline doesn't appear to really abide by this rule, aside from the very edge of the light area.

**Pull request merged**